xtask/src/cargo.rs

```diff
@@ -0,0 +1,211 @@
+use std::path::PathBuf;
+
+use isahc::{HttpClient, ReadResponseExt};
+use semver::{Version, VersionReq};
+use serde::{de::IgnoredAny, Deserialize};
+use serde_json::from_str as from_json_str;
+use toml_edit::{value, Document};
+use xshell::{cmd, pushd, read_file, write_file};
+
+use crate::{util::ask_yes_no, Result};
+
+const CRATESIO_API: &str = "https://crates.io/api/v1/crates";
+
+/// The metadata of a cargo workspace.
+#[derive(Clone, Debug, Deserialize)]
+pub struct Metadata {
+    pub workspace_root: PathBuf,
+    pub packages: Vec<Package>,
+}
+
+impl Metadata {
+    /// Load a new `Metadata` from the command line.
+    pub fn load() -> Result<Metadata> {
+        let metadata_json = cmd!("cargo metadata --no-deps --format-version 1").read()?;
+        Ok(from_json_str(&metadata_json)?)
+    }
+}
+
+/// A cargo package.
+#[derive(Clone, Debug, Deserialize)]
+pub struct Package {
+    /// The package name
+    pub name: String,
+
+    /// The package version.
+    pub version: Version,
+
+    /// The package's manifest path.
+    pub manifest_path: PathBuf,
+
+    /// A map of the package dependencies.
+    #[serde(default)]
+    pub dependencies: Vec<Dependency>,
+}
+
+impl Package {
+    /// Update the version of this crate.
+    pub fn update_version(&mut self, version: &Version) -> Result<()> {
+        println!("Updating {} to version {}…", self.name, version);
+
+        let mut document = read_file(&self.manifest_path)?.parse::<Document>()?;
+
+        document["package"]["version"] = value(version.to_string());
+
+        write_file(&self.manifest_path, document.to_string())?;
+
+        self.version = version.clone();
+
+        Ok(())
+    }
+
+    /// Update the version of this crate in dependant crates' manifest, with the given version
+    /// prefix.
+    pub fn update_dependants(&self, packages: &[Package]) -> Result<()> {
+        for package in
+            packages.iter().filter(|p| p.dependencies.iter().any(|d| d.name == self.name))
+        {
+            println!("Updating dependency in {} crate…", package.name);
+
+            let mut document = read_file(&package.manifest_path)?.parse::<Document>()?;
+
+            for dependency in package.dependencies.iter().filter(|d| d.name == self.name) {
+                let version = if dependency.req.is_exact() {
+                    format!("={}", self.version,)
+                } else {
+                    self.version.to_string()
+                };
+
+                match dependency.kind {
+                    Some(DependencyKind::Dev) => {
+                        document["dev-dependencies"][&self.name]["version"] =
+                            value(version.as_str());
+                    }
+                    None => {
+                        document["dependencies"][&self.name]["version"] = value(version.as_str());
+                    }
+                    _ => {}
```

Comment:
We don't currently have any build-deps, but maybe we should still include the necessary few lines to support that. Then you could also just return the name of the toml section from the match and do the assignment below.